### DISCUSSION TITLE
Importing an npm module into a notebook?

### DISCUSSION THREAD

**Discussion by kolibril13** [rgbkrk/denotebooks#4] [Q&A]
hi @rgbkrk,
I was wondering: the docs say https://docs.deno.com/runtime/manual/tools/jupyter
<img width="813" alt="image" src="https://github.com/rgbkrk/denotebooks/assets/44469195/539e4847-6503-451c-9fa7-7bb3231ca95c">

How can I do this import from npm?
E.g. I'd like to import this module: https://www.npmjs.com/package/@niivue/niivue

I've already made a python widget from niivue that using anywidget, but would be interesting to see that working in a deno kernel as well:

https://github.com/rgbkrk/denotebooks/assets/44469195/9a54e8f4-33e1-49bf-accf-06a4d3e4f763

this tweet mentioned a few months ago that something like this might be possible

https://twitter.com/KyleRayKelley/status/1708319670535877070
<img width="604" alt="image" src="https://github.com/rgbkrk/denotebooks/assets/44469195/22594408-5dc7-4c45-aee1-2aa6f58d7718">

is there an example somewhere that you can point me to?
Thanks a lot!

**Comment by bartlomieju** [comment]
You can import it like so:
```js
import niivue from "npm:@niivue/niivue";
```

**Comment by kolibril13** [comment]
thanks for the feedback!
I've tried
<img width="1046" alt="image" src="https://github.com/rgbkrk/denotebooks/assets/44469195/45e86c2b-8a66-4ded-b29e-e1fef2c97a16">
but that does not work.
Example from here:
https://github.com/niivue/niivue/blob/main/DEVELOP.md#load-a-volume
When I run the same code using vite+react, it does work:
<img width="1313" alt="image" src="https://github.com/rgbkrk/denotebooks/assets/44469195/35cff623-7f9c-4dda-aef0-0c0e623cc318">
any ideas? :)

**Comment by bartlomieju** [comment]
This is probably because Deno has `window` and `devicePixelRatio` is not supported. Feel free to open an issue in Deno repo?

**Comment by rgbkrk** [comment]
I'd need to play with niivue to know for sure, but I think this relates to https://github.com/denoland/deno/issues/20592. We need to support comms in order to have bidirectional communication with a widget like this and the package needs to be rendered on the client, not on the Deno side.

An initial prototype way to use anywidget in Deno is in https://deno.land/x/anywidget@0.0.5

**Comment by manzt** [comment]
For clarity, Jupyter Widgets connect a Jupyter front-end (VSCode, browser) with objects in the Jupyter kernel (Python, R, Deno, ...). The Deno integration implements a _kernel_. However, JS libraries that are intended to run in client (i.e., most interactive visualization libraries that use Canvas API or WebGL) still need to run in the _client_ (unless Deno implements these WebAPIs like @bartlomieju mentioned, but such implementations probably wouldn't be what is desired in this context of an interactive Jupyter notebooks). The Deno kernel lets you bridge client/kernel more conveniently because it's the same language/stack, but you need to be mindful of what packages are meant to run where.

You could take your existing client bundle (`static/widget.js`) for your Python ipyniivue example, and expose it to the Deno kernel with https://deno.land/x/anywidget@0.0.5 like @rgbkrk mentioned. FWIW, I think there is a lot of room to improve the ergonomics of making Deno-based Jupyter widgets, building upon the rich full-stack frameworks available in the Deno ecosystem (e.g. widgets are kind of like "islands" in fresh https://deno.com/blog/intro-to-islands).
